### DISCUSSION TITLE
How to make 2 Quasar toggle button groups mutually exclusive?

### DISCUSSION THREAD

**Discussion by tesicg** [quasarframework/quasar#10663] [General - Components / Directives / etc] [answered]
We have Vue.js app that uses Quasar component framework.

The screen shots look as following:

Incorrect case:

[![enter image description here][1]][1]

Correct case:

[![enter image description here][2]][2]

I need the only one toggle button group should be active, either percents or standard amounts. ```Pay attention there's an array of toggle button groups.```

The code I wrote produces the incorrect case. It looks as following:

    <template>
      <q-dialog v-model="show" no-backdrop-dismiss full-width>
        <q-card>
          <q-card-section class="row items-center">
            <div class="text-h6 tip-color">Tip</div>
            <q-space />
            <q-btn icon="close" flat round dense v-close-popup @click="cancel" />
          </q-card-section>
          <q-card-section style="max-height: 50vh" class="scroll set-border">
            <div class="q-gutter-md">
              <q-card v-for="({ master }, index) in items" :key="master.id">
                <div class="row">
                  <div class="col-3 flex justify-center items-center">
                    <Avatar
                      :src="master.employer_avatar"
                      :size="50"
                      no-default-spinner
                    />
                    <span class="q-ml-md text-caption text-secondary">
                      {{ getEmployerName(master) }}
                    </span>
                  </div>
                  <div class="col">
                    <q-card-section>
                      <q-btn-toggle
                        v-model="togglePercentPayments[index]"
                        toggle-color="primary"
                        :options="percentPayments"
                        spread
                        @input="getFullTip"
                      />
                    </q-card-section>
                    <q-card-section>
                      <q-btn-toggle
                        v-model="toggleStandardPayments[index]"
                        toggle-color="primary"
                        :options="standardPayments"
                        spread
                        @input="getFullTip"
                      />
                    </q-card-section>
                  </div>
                </div>
              </q-card>
            </div>
          </q-card-section>
          <q-card-section>
            <div><strong>Summary:</strong> {{ total }}</div>
            <div>
              <strong>Tip:</strong> {{ getFullTip() ? getFullTip() : 0 }}
            </div>
          </q-card-section>
          <q-card-actions align="right" class="text-primary q-pt-none">
            <q-btn flat label="Cancel" @click="cancel" />
            <q-btn flat label="Pay" @click="pay" />
          </q-card-actions>
        </q-card>
      </q-dialog>
    </template>

    <script>
    const percentPayments = [
      { label: '5%', value: 5 },
      { label: '10%', value: 10 },
      { label: '15%', value: 15 },
    ]

    const standardPayments = [
      { label: '100', value: 100 },
      { label: '200', value: 200 },
      { label: '500', value: 500 },
    ]

    export default {
      props: {
        showModal: {
          type: Boolean,
          default: false,
        },
        items: {
          type: Array,
          default: () => [],
        },
        total: {
          type: Number,
          default: 0,
        },
      },
      data: function() {
        return {
          show: false,
          togglePercentPayments: new Array(this.items.length).fill(null),
          toggleStandardPayments: new Array(this.items.length).fill(null),
          percentPayments,
          standardPayments,
        }
      },
      watch: {
        showModal(newVal) {
          this.show = newVal
        },
      },
      methods: {
        getEmployerName(master) {
          return `${master?.first_name ?? ''} ${master?.last_name[0] ?? ''}.`
        },
        getPercentage(total, percent) {
          return (total / 100) * percent
        },
        getFullTip() {
          let standardSum = 0
          for (const standardPayment of this.toggleStandardPayments) {
            standardSum += standardPayment
          }

          let percentageSum = 0
          for (const percentagePayment of this.togglePercentPayments) {
            const percent = this.getPercentage(this.total, percentagePayment)
            percentageSum += percent
          }

          return standardSum + percentageSum
        },
        pay() {
          this.$emit('pay', this.getFullTip())
          this.clear()
        },
        cancel() {
          this.$emit('cancel')
          this.clear()
        },
        clear() {
          this.togglePercentPayments = new Array(this.items.length).fill(null)
          this.toggleStandardPayments = new Array(this.items.length).fill(null)
        },
      },
    }
    </script>

    <style scoped>
    .set-border {
      border: 1px solid gainsboro;
    }
    .tip-color {
      color: rgb(4, 171, 171);
    }
    </style>

How to make 2 Quasar toggle button groups mutually exclusive?

  [1]: https://i.stack.imgur.com/o2XoR.png
  [2]: https://i.stack.imgur.com/lzEm1.png

**Comment by ontwikkelfabriek** [comment]
@tesicg 

Please create a codepen.io that demo's your problem instead of posting a long code listing ( use JS highlighting ). So  we can provide better support to you and others with similar questions.

**Reply by tesicg** [reply]
I've never used codepen to create something before. I've registered for the first time and tried Vue template, but there is no way to add new file (I can't see sidebar). Also, I've tried to add the links to Quasar css and js, but it doesn't work. I've added the following:

https://fonts.googleapis.com/css?family=Roboto:100,300,400,500,700,900|Material+Icons
https://cdn.jsdelivr.net/npm/quasar@2.0.4/dist/quasar.prod.css

https://cdn.jsdelivr.net/npm/quasar@2.0.4/dist/quasar.umd.prod.js

What should I do to add the links to Quasar css and js?

Should I add the links to Vue libraries?

**Reply by smolinari** [reply]
@tesicg - The mutually exclusive part is relatively simple. Just use one model for each card, and for each toggle button set, one set of options but with different values (I don't have cards in the codepen, but from what is there, you can get the picture). The values can and should be different of course and the selected value should be recalculated to get you the final usable value. 

https://codepen.io/smolinari/pen/vYZePqZ

Scott

p.s. btw, I had some time on my hands. But for sure, you should ask with a prebuilt codepen showing your current logic, as @ontwikkelfabriek noted.

**Reply by tesicg** [reply]
I can't use hardcoded models. I have an array as a model. That's the issue.

**Reply by smolinari** [reply]
@tesicg - You mean like this?

https://codepen.io/smolinari/pen/gORGJby?editors=1010

Scott

**Comment by tesicg** [comment]
I'm trying to set my codepen up. Here's what I have:

HTML:

```
<div id="q-app">
	<q-card>
		<q-card-section class="scroll set-border">
			<q-card v-for="(master, index) in testItems" :key="master.id">
				<q-card-section>
					<q-btn-toggle
						v-model="togglePercentPayments[index]"
						toggle-color="primary"
						:options="percentPayments"
						spread
						@input="getFullTip"
					/>
				</q-card-section>
				<q-card-section>
					<q-btn-toggle
						v-model="toggleStandardPayments[index]"
						toggle-color="primary"
						:options="standardPayments"
						spread
						@input="getFullTip"
					/>
				</q-card-section>
			</q-card>
		</q-card-section>
		<q-card-section>
			<div><strong>Summary:</strong> {{ total }}</div>
			<div>
				<strong>Tip:</strong> {{ getFullTip() ? getFullTip() : 0 }}
			</div>
		</q-card-section>
	</q-card>
</div>
```

CSS:

```
.set-border {
  border: 1px solid gainsboro;
}
.tip-color {
  color: rgb(4, 171, 171);
}
```

JS:

```
new Vue({
  el: '#q-app',
  data: function () {
    return {
      testItems: [
        { id: 1, name: 'Elena' },
        { id: 2, name: 'Alona' },
        { id: 3, name: 'Irina' },
        { id: 4, name: 'Tanya' },
        { id: 5, name: 'Anna' },
      ],
      percentPayments: [
        { label: '5%', value: 5 },
        { label: '10%', value: 10 },
        { label: '15%', value: 15 },
      ],
      standardPayments: [
        { label: '100', value: 100 },
        { label: '200', value: 200 },
        { label: '500', value: 500 },
      ],
      togglePercentPayments: new Array(this.testItems.length).fill(null),
      toggleStandardPayments: new Array(this.testItems.length).fill(null),
      percentPayments,
      standardPayments,
    }
  },
  methods: {
    getPercentage: function (total, percent) {
      return (total / 100) * percent
    },
    getFullTip: function () {
      let standardSum = 0
      for (const standardPayment of this.toggleStandardPayments) {
        standardSum += standardPayment
      }

      let percentageSum = 0
      for (const percentagePayment of this.togglePercentPayments) {
        const percent = this.getPercentage(this.total, percentagePayment)
        percentageSum += percent
      }

      return standardSum + percentageSum
    },
  }
})
```

But, console (F12) shows an error:

```
TypeError: Cannot read properties of undefined (reading 'length')
```

**Comment by tesicg** [comment]
The less original code:

```
<template>
  <q-dialog v-model="show" no-backdrop-dismiss full-width>
    <q-card>
      <q-card-section style="max-height: 50vh" class="scroll set-border">
        <div class="q-gutter-md">
          <q-card v-for="({ master }, index) in items" :key="master.id">
            <div class="row">
              <div class="col">
                <q-card-section>
                  <q-btn-toggle
                    v-model="togglePercentPayments[index]"
                    toggle-color="primary"
                    :options="percentPayments"
                    spread
                    @input="getFullTip"
                  />
                </q-card-section>
                <q-card-section>
                  <q-btn-toggle
                    v-model="toggleStandardPayments[index]"
                    toggle-color="primary"
                    :options="standardPayments"
                    spread
                    @input="getFullTip"
                  />
                </q-card-section>
              </div>
            </div>
          </q-card>
        </div>
      </q-card-section>
      <q-card-section>
        <div><strong>Summary:</strong> {{ total }} руб.</div>
        <div>
          <strong>Tip:</strong> {{ getFullTip() ? getFullTip() : 0 }} руб.
        </div>
      </q-card-section>
    </q-card>
  </q-dialog>
</template>

<script>
const percentPayments = [
  { label: '5%', value: 5 },
  { label: '10%', value: 10 },
  { label: '15%', value: 15 },
]

const standardPayments = [
  { label: '100', value: 100 },
  { label: '200', value: 200 },
  { label: '500', value: 500 },
]

export default {
  props: {
    showModal: {
      type: Boolean,
      default: false,
    },
    items: {
      type: Array,
      default: () => [],
    },
    total: {
      type: Number,
      default: 0,
    },
  },
  data: function() {
    return {
      show: false,
      togglePercentPayments: new Array(this.items.length).fill(null),
      toggleStandardPayments: new Array(this.items.length).fill(null),
      percentPayments,
      standardPayments,
    }
  },
  watch: {
    showModal(newVal) {
      this.show = newVal
    },
  },
  methods: {
    getPercentage(total, percent) {
      return (total / 100) * percent
    },
    getFullTip() {
      let standardSum = 0
      for (const standardPayment of this.toggleStandardPayments) {
        standardSum += standardPayment
      }

      let percentageSum = 0
      for (const percentagePayment of this.togglePercentPayments) {
        const percent = this.getPercentage(this.total, percentagePayment)
        percentageSum += percent
      }

      return standardSum + percentageSum
    },
  },
}
</script>

<style scoped>
.set-border {
  border: 1px solid gainsboro;
}
.tip-color {
  color: rgb(4, 171, 171);
}
</style>
```

**Reply by smolinari** [reply]
@tesicg - Did you see my second codepen? 

https://codepen.io/smolinari/pen/gORGJby?editors=1010

Scott

**Reply by tesicg** [reply]
It helped. I've solved my problem. Thank you.

How can I check this answer (with the second codepen example) as the right one?

**Comment by smolinari** [comment, marked as answer]
@tesicg - Did you see my second codepen?

https://codepen.io/smolinari/pen/gORGJby?editors=1010

Scott
